src/System.Private.CoreLib/shared/System/Number.NumberToFloatingPointBits.cs

```diff
@@ -0,0 +1,627 @@
+
+            fractionalNumerator.ShiftLeft(remainingBitsOfPrecisionRequired);
+
+            BigInteger.DivRem(ref fractionalNumerator, ref fractionalDenominator, out BigInteger bigFractionalMantissa, out BigInteger fractionalRemainder);
+            ulong fractionalMantissa = bigFractionalMantissa.ToUInt64();
+            bool hasZeroTail = !number.HasNonZeroTail && fractionalRemainder.IsZero();
+
+            // We may have produced more bits of precision than were required.  Check,
+            // and remove any "extra" bits:
+            uint fractionalMantissaBits = BigInteger.CountSignificantBits(fractionalMantissa);
+
+            if (fractionalMantissaBits > requiredFractionalBitsOfPrecision)
+            {
+                int shift = (int)(fractionalMantissaBits - requiredFractionalBitsOfPrecision);
+                hasZeroTail = hasZeroTail && (fractionalMantissa & ((1UL << shift) - 1)) == 0;
+                fractionalMantissa >>= shift;
+            }
+
+            // Compose the mantissa from the integer and fractional parts:
+            ulong integerMantissa = integerValue.ToUInt64();
+            ulong completeMantissa = (integerMantissa << (int)(requiredFractionalBitsOfPrecision)) + fractionalMantissa;
+
+            // Compute the final exponent:
+            // * If the mantissa had an integer part, then the exponent is one less than
+            //   the number of bits we obtained from the integer part.  (It's one less
+            //   because we are converting to the form 1.11111, with one 1 to the left
+            //   of the decimal point.)
+            // * If the mantissa had no integer part, then the exponent is the fractional
+            //   exponent that we computed.
+            // Then, in both cases, we subtract an additional one from the exponent, to
+            // account for the fact that we've generated an extra bit of precision, for
+            // use in rounding.
+            int finalExponent = (integerBitsOfPrecision > 0) ? (int)(integerBitsOfPrecision) - 2 : -(int)(fractionalExponent) - 1;
+
+            return AssembleFloatingPointBits(in info, completeMantissa, finalExponent, hasZeroTail);
+        }
+
+        private static ulong RightShiftWithRounding(ulong value, int shift, bool hasZeroTail)
+        {
+            // If we'd need to shift further than it is possible to shift, the answer
+            // is always zero:
+            if (shift >= 64)
+            {
+                return 0;
+            }
+
+            ulong extraBitsMask = (1UL << (shift - 1)) - 1;
+            ulong roundBitMask = (1UL << (shift - 1));
+            ulong lsbBitMask = 1UL << shift;
+
+            bool lsbBit = (value & lsbBitMask) != 0;
+            bool roundBit = (value & roundBitMask) != 0;
+            bool hasTailBits = !hasZeroTail || (value & extraBitsMask) != 0;
+
+            return (value >> shift) + (ShouldRoundUp(lsbBit: lsbBit, roundBit: roundBit, hasTailBits: hasTailBits) ? 1UL : 0);
```

Comment:
(nit) the method call here is a bit redundant in the argument passing, which makes it hard to read IMO:

`ShouldRoundUp(lsbBit: lsbBit, roundBit: roundBit, hasTailBits: hasTailBits)` =>
`ShouldRoundUp(lsbBit, roundBit, hasTailBits)`

Reply:
Fixed.